text/0096-rawkv-stale-read.md

```diff
@@ -0,0 +1,97 @@
+
+    Lease use clocks instead of messages to handle read-only queries, the leader would use the normal heartbeat mechanism to maintain a lease (usually shorter than election timeout for the clock drift). This mechanism could improve the read latency dramatically. However, it still requires the client to read from the leader instead of other replicas. If the client and the leader are located in different regions, the read latency could be terribly high.
+
+    <p align="center">
+        <img src="../media/lease.png" width="450px"/>
+    </p>
+
+3. Stale read for TxnKV only.
+
+    Stale read on TxnKV allows to read outdated data from the replica other than the leader but still keeping snapshot isolation provided by TiKV transaction. Because the client tells the follower the timestamp of the stale snapshot it wants to read on (rather than read-the-latest), the follower can check if the data versions that are fresh enough via the `saft_ts` (the minimum timestamp of the on-the-fly transaction). Any read transaction with a smaller timestamp than `safe_ts` could be preserved the snapshot isolation.
+
+    <p align="center">
+        <img src="../media/txnkv-stale-read.svg" width="450px"/>
+    </p>
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before (read committed).
+
+The follower will read the local storage with a `read_ts` that reuses the mechanism from the stale read of TxnKV. Surprisingly, the leader can also stale read, and in this case, the leader can read locally without a lease.
+
+Note that the freshness of data may regress from the point of view of the client since the client may choose different followers to read from time to time.
+
+<p align="center">
+    <img src="../media/rawkv-stale-read.svg" width="450px"/>
+</p>
+
+
+## Detailed Design
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy time-bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before.
+
+### TiKV
+
+ While trying to read data, clients should specify a timestamp which attachs to the request header as `read_ts`, typically a timestamp few seconds ago. The replica should read the local storage with the `read_ts` and reuses the mechanism from the stale read of TxnKV. This requires the replica to check the `read_ts` against the `safe_ts` which is advaneced by `CheckLeader` message from the store of the leader or `resolve-ts` worker. As long as the `safe_ts` is no less than `read_ts`, the replica is allowed to read the key from local storage.
```

Comment:
How is `safe_ts` maintained in RawKV since there is no locks?

Reply:
If there are no locks, the `resolve_ts` will advance the `safe_ts` by requesting the TSO for a timestamp periodically. The default config for the `resolve_ts` worker is `1s`.

Reply:
More details are supplemented.